integration-tests/src/test/scala/chiselTest/util/SparseVecSpec.scala

```diff
@@ -0,0 +1,356 @@
+  mapping:              Seq[(Int, UInt)],
+  expected:             Seq[(Int, Data)],
+  debug:                Boolean = false)
+    extends BasicTester {
+  // Create a wire SparseVec and initialize it to the values in the mapping.
+  private val sparseVec = Wire(new SparseVec(size, tpe, mapping.map(_._1), defaultValueBehavior, outOfBoundsBehavior))
+  sparseVec.elements.values.zip(mapping.map(_._2)).foreach { case (a, b) => a :<>= b }
+
+  class TestBundle extends Bundle {
+    val index = UInt()
+    val value = UInt()
+  }
+
+  val tests = Wire(
+    Vec(
+      expected.size,
+      new TestBundle
+    )
+  )
+  expected.zipWithIndex.foreach {
+    case ((index, value), testNumber) =>
+      tests(testNumber).index := index.U
+      tests(testNumber).value := value
+  }
+
+  // Access the dense vector and the sparse vector, using all of the access
+  // types, and make sure that the results are exactly the same.
+  private val (index, wrap) = Counter(0 until tests.size)
+  private val failed = RegInit(Bool(), false.B)
+  private val reference = tests(index)
+  private val sparseVecResults = Seq(Lookup.Binary, Lookup.OneHot, Lookup.IfElse).map(sparseVec(reference.index, _))
+  if (debug)
+    printf(
+      "%x: %x, %x, %x, %x",
+      reference.index,
+      reference.value,
+      sparseVecResults(0),
+      sparseVecResults(1),
+      sparseVecResults(2)
+    )
+  when(sparseVecResults.map(_ =/= reference.value).reduce(_ || _)) {
+    failed := true.B
+    if (debug)
+      printf(" <-- error")
+    else
+      assert(false.B)
+  }
+  if (debug)
+    printf("\n")
+
+  when(wrap) {
+    when(RegNext(true.B)) {
+      stop()
+    }
+  }
+
+}
+
+class SparseVecSpec extends ChiselFlatSpec with Utils {
+  "SparseVec equivalence to Dynamic Index" should "work for a complete user-specified mapping" in {
+    assertTesterPasses(
+      new SparseVecDynamicIndexEquivalenceTest(
+        4,
+        UInt(3.W),
+        Seq(
+          0 -> 1.U,
+          1 -> 2.U,
+          2 -> 3.U,
+          3 -> 4.U
+        )
+      )
+    )
+  }
+
+  it should "work for a mapping that includes default values" in {
+    assertTesterPasses(
+      new SparseVecDynamicIndexEquivalenceTest(
+        4,
+        UInt(3.W),
+        Seq(
+          0 -> 1.U,
+          1 -> 2.U,
+          3 -> 4.U
+        )
+      )
+    )
+  }
+
+  it should "work for a mapping that includes out-of-bounds accesses" in {
+    assertTesterPasses(
+      new SparseVecDynamicIndexEquivalenceTest(
+        3,
+        UInt(3.W),
+        Seq(
+          0 -> 1.U,
+          1 -> 2.U,
+          2 -> 3.U
+        )
+      )
+    )
+  }
+
+  it should "work for a mapping that includes out-of-bounds accesses and no zeroth element" in {
+    assertTesterPasses(
+      new SparseVecDynamicIndexEquivalenceTest(
+        3,
+        UInt(3.W),
+        Seq(
+          1 -> 2.U,
+          2 -> 3.U
+        ),
+        true
+      )
+    )
+  }
+
```

Comment:
Doesn't the size need to be at least 4 for this to have out-of-bounds access?

Reply:
The _access_ needs to be padded up to a power of 2, yes. However, the size of the vec should be smaller. The test checks all values up to the next power of 2.

This test is doing is constructing a `SparseVec` and a dense `Vec` of size 3. The dense `Vec` has all values initialized to `DontCare` (elements `[0, 1, 2]`). The user-provided indices in both the `Vec` and `SparseVec` are then set (`[1, 2]`). The test then checks that all indices, including out-of-bounds indices match. I.e., all of `[0, 1, 2, 3]` match.

There's a `debug` flag to the test that shows the index/output behavior for this test. Check if this looks right to you, specifically, `0 -> 0` and `3 -> 0`:

```
index, dense, binary, onehot, ifelse
0: 0, 0, 0, 0
1: 2, 2, 2, 2
2: 3, 3, 3, 3
3: 0, 0, 0, 0
```

Reply:
I added some additional comments to the tests to indicate what is going on here.